### ISSUE TITLE
[Inductor] support complex dtypes for randn 

### ISSUE THREAD

**Issue by min-jean-cho** [pytorch/pytorch#92398]
### 🚀 The feature, motivation and pitch

I see Inductor cpp codgegen `DTYPE_TO_CPP`, `DTYPE_TO_ATEN` do not support complex dtypes.
https://github.com/pytorch/pytorch/blob/9b173b87b2587ee085aaef65e1a0fdebd90d5229/torch/_inductor/codegen/cpp.py#L31-L55

 
For example, with `randn` (complex dtype support is missing for other ops as well (e.g., `rand`, `sqrt`, etc))
```python
import torch

def foo(x):
    return torch.randn(x.size(), dtype=x.dtype)

opt_foo = torch.compile(foo)
print("eager: ", foo(torch.empty(1, dtype=torch.complex64)))
print("inductor: ", opt_foo(torch.empty(1, dtype=torch.complex64))) 
```
raises the following error from codegen
```
return f"static_cast<{DTYPE_TO_CPP[dtype]}>(randn_cpu({seed}, {offset}));"
KeyError: 'torch.complex64\n\nWhile executing %randn : [#users=1] = call_method[target=randn](args = (%ops, %bitwise_xor, %constant_1, torch.complex64), kwargs = {})\nOriginal traceback:\nNone'
```

I wonder if there's any context to why complex dtypes are currently generally not supported in Inductor.

For `randn`, we can extend `Philox4_32` `randn` to support complex dtypes.
```
Z ~ Complex Normal(0, 1)
Z = 1/sqrt(2) * [X + Xj],     X ~ Normal(0, 1) 
```

**Comment by min-jean-cho** [comment]
@jgong5, @ashokei, please help let me know if any comments on this.

**Comment by jgong5** [comment]
yes, complex dtype is not supported yet. You may have to add `c10::complex` to the table. It should have most of the operator overloads needed for codegen. Welcome to submit a PR on it.

**Comment by min-jean-cho** [comment]
Thanks @jgong5, sounds good to me.

**Comment by lezcano** [comment]
Note that if we want to start supporting complex dtypes we'd need to go all the way and start supporting them all throughout, not for just for `randn`. I'm changing the title of the issue accordingly.

**Comment by mfkasim1** [comment]
@min-jean-cho Is there any update on this? I can help if you want to divide the task

**Comment by min-jean-cho** [comment]
@mfkasim1, feel free to take this task if you would like! I haven't had the time to work on this.

**Comment by khushi-411** [comment]
Hi, all!
I see @mfkasim1 mentioned https://github.com/pytorch/pytorch/pull/95484#issuecomment-1494008457. I'd like to pick this up! Working on this seems fun, and there are many new concepts to learn.

I'm just writing here for visibility. Please let me know if @mfkasim1 or anyone else has started working on this. Thanks :-)

**Comment by lezcano** [comment]
This is a rather difficult task with *many* edge cases, and in the best case scenario, it would be just possible to add support for CPU, as triton does not support cuda yet.

Now, if you think that you could submit a PR adding complex support for CPU, by all means go for it, but I don't think anyone in the inductor currently has the bandwidth to assist on this issue.

Note that the script on the OP already correctly fallsback to eager, so at least there is no correctness issue there.

**Comment by khushi-411** [comment]
Thanks, @lezcano, for the heads-up! :-)
Sure, I'd like to try this on the CPU side.

**Comment by hlc1209** [comment]
It would be fantastic if complex numbers were supported. Many disciplines, including but not limited to physics and engineering, are eagerly anticipating this feature. Moreover, the Fourier transform is increasingly applied in computer vision powered by deep learning

**Comment by zaptrem** [comment]
I've run into this error while writing a transformer with rotary embeddings (RoPE). Since this thread hasn't been spammed with similar issues I assume there's a standard way to work around it. Would any of you be able to share that?

**Comment by lezcano** [comment]
A way to implement rotatory embeddings: https://github.com/pytorch-labs/gpt-fast/blob/8c8c463f1f909890a2c0cd3f9e2ad75fff7f2b6c/model.py#L226-L238. More generally, these questions are better suited for the forums: https://discuss.pytorch.org/

**Comment by jiayisunx** [comment]
This PR(https://github.com/pytorch/pytorch/pull/103183) already supports complex64 in Inductor cpp codgegen. I guess it should have most of the operator overloads needed for codegen. And I verified that this randn example works fine.

**Comment by lezcano** [comment]
Complex types are supported as a blackbox by falling back to eager. The problem with this approach is that this completely deactivates fusion for these ops. This issue proposes supporting complex numbers natively within inductor.

**Comment by ezyang** [comment]
So, to be clear, the agreed upon plan for CUDA complex support is that we are going to we are going to replace native complex tensors with a subclass that keeps real and imag as separate tensors https://github.com/albanD/subclass_zoo/pull/48 and then use @bdhirsh tensor subclass compilation to compile it into native code without any "extra" complex support for inductor.

On CPU you are welcome to add std::complex codegen if you really want to, but on CUDA, to get fast matmuls we need to keep real/imag separate.

**Comment by jgong5** [comment]
> On CPU you are welcome to add std::complex codegen if you really want to, but on CUDA, to get fast matmuls we need to keep real/imag separate.

Hmm, I don't see a reason why CPU has to be different. I guess this can be shared among devices?

**Comment by amjames** [comment]
I was not tracking this issue in relation to the work in #169832 I am noticing a decent number of PRs looking to add complex specific lowering/decomps and I am trying to clean them up. 

The plan detailed in [this comment](https://github.com/pytorch/pytorch/issues/92398#issuecomment-1883336727) has been implemented  in #169832, but it was not limited specifically to `cuda` devices. That could be changed if there is interest and strong motivation for using native std::complex codegen for CPU in the future. 

I am going to close this as it is no longer necessary to support the codegen mapping for complex dtypes since we are always working with real tensors by the time we reach the codegen layer.
